evap/evaluation/models.py

```diff
@@ -1,34 +1,332 @@
+
+        field_name = next((field.name for field in type(instance)._meta.many_to_many
+                           if getattr(type(instance), field.name).through == sender), None)
+
+        if action == 'pre_remove':
+            instance._m2m_changes[field_name]['remove'] += list(pk_set)
+        elif action == 'pre_add':
+            instance._m2m_changes[field_name]['add'] += list(pk_set)
+        elif action == 'pre_clear':
+            instance._m2m_changes[field_name]['clear'] = []
+
+        if "pre" in action:
+            instance.update_log("m2m")
+
+    def _as_dict(self, include_m2m=False):
+        """
+        Return a dict mapping field names to values saved in this instance.
+        Only include field names that are not to be ignored for logging.
+        Except when deleting objects, m2m values come from signal handling.
+        """
+        fields = [
+            field.name for field in type(self)._meta.get_fields() if
+            field.name not in self.ignore_field_names_logging
+            and (include_m2m or not field.many_to_many)
+        ]
+        return model_to_dict(self, fields)
+
+    def _change_data(self, action_type, include_none_values=False):
+        """
+        Return a dict mapping field names to changes that happened in this model instance,
+        depending on the action that is being done to the instance.
+        """
+        self_dict = self._as_dict()
+        if action_type == "create":
+            changes = {
+                field_name: {'change': [None, created_value]}
+                for field_name, created_value in self_dict.items()
+                if created_value is not None or include_none_values
+            }
+        elif action_type == "change":
+            old_dict = type(self).objects.get(pk=self.pk)._as_dict()
+            changes = {
+                field_name: {'change': [old_value, self_dict[field_name]]}
+                for field_name, old_value in old_dict.items()
+                if old_value != self_dict[field_name]
+            }
+        elif action_type == "delete":
+            old_dict = type(self).objects.get(pk=self.pk)._as_dict(include_m2m=True)
+            changes = {}
+            for field_name, old_value in old_dict.items():
+                if old_value is None and not include_none_values:
+                    continue
+                field = self._meta.get_field(field_name)
+                if field.many_to_many:
+                    action_items = [obj.pk for obj in old_value]
+                else:
+                    action_items = [old_value]
+                changes[field_name] = {'delete': action_items}
+        else:
+            raise ValueError("Unknown action type: '{}'".format(action_type))
+
+        changes.update(self._m2m_changes)
+        return changes
+
+    def update_log(self, mode):
+        action = {
+            'delete': 'delete',
+            'create': 'create',
+            'change': 'change',
+            'm2m': 'change',
```

Comment:
right now there are these three sets of strings as far as i can see: delete|create|change|m2m (modes?), delete|create|change (actions?), and add|remove|clear (some other actions?). clearly distinguishing between these three sets with some prefixes would be nice, and using constants instead of magic strings too. enums might be useful here.

Reply:
I created enums, but I'm unsure about how to access them in the template. Using a context processor providing the two enums looks like a valid option. Do you have another idea?